### PR TITLE
Windows CLI Ansi Rewrite

### PR DESCRIPTION
**Pull request by gabrielhartmann**
A large portion of the Windows CLI has been rewritten.  Ansi parsing and the actions taken as a result of Ansi commands are now decoupled.  There have also been a number of bug fixes in the actions taken in response to Ansi commands in the Windows environment.

Vi now functions as expected and the vttest no longer panics.  Further improvement towards full Ansi terminal emulator compliance is needed, but much of the standard most commonly used features are now properly supported.

### CODE REVIEW

**Comment by jhowardmsft** [comment]
Cool! :)

**Comment by crosbymichael** [comment]
ping @icecrime

**Comment by ahmetalpbalkan** [comment]
@crosbymichael this is blocked on #13257, we will come back to this. Please ignore this PR for now.

File: hack/vendor.sh [hunk 1]
```diff
@@ -59,6 +59,8 @@ clone git github.com/go-fsnotify/fsnotify v1.2.0
 
 clone git github.com/go-check/check 64131543e7896d5bcc6bd5a76287eb75ea96c673
 
+clone git github.com/Azure/go-ansiterm
```
Comment (LK4D4) [inline hack/vendor.sh]: Wow, this is awesome :)

Comment (gabrielhartmann) [inline hack/vendor.sh]: It's a much smaller pull request now.  And in the mean time I've fixed all known bugs.  There are no panics or odd behavior in any known real world scenario.  The synthetic test vttest doesn't pass, but iTerm doesn't pass that test either.

**Comment by ahmetalpbalkan** [comment]
Hi @icecrime @tiborvass @crosbymichael,

Gabriel has just updated this pull request. We are now vendoring http://github.com/Azure/go-ansiterm package and the docker codebase only has a little piece of integration code. Please review and let @gabrielhartmann know what you think. :smile:

**Comment by tiborvass** [comment]
Thanks guys!

**Comment by jhowardmsft** [comment]
This is a lot better. A LOT better!. There are still some bugs though. Screenshots below. I tried docker pull Ubuntu:15.04 from Linux (against a Linux daemon). Compare with the output on Windows doing the same, targetting both the same Linux daemon, and a Windows daemon for good measure ( ;) ). Note how the Windows output is not quite correct. Using 1.7 master with the commits from this PR cherry-picked on Windows. On Linux, using 1.7 master.
ansi from linux
ansi from windows

**Comment by ahmetalpbalkan** [comment]
@jhowardmsft apparently it's hard to take a diff by looking at 3 things by eye. what's the difference between these 3?

**Comment by jhowardmsft** [comment]
Notice how in Windows, no gap between "15.04 Pulling from Ubuntu" and the next lines. Also notice how c6a row appears twice in the first picture. And how the result of the "second" c6a is still in-progress. You can get the same by pointing the Windows CLI at Ubuntu - just the graph layers weren't consistent between the Linux and Windows daemons. Hence I know it's not a Windows daemon issue. Should have done a better screen-shot ;) If it helps, it's easier to see while the pull is in progress. Here's a sample of a Windows client pointing at a Linux daemon during the pull (after a docker rmi Ubuntu:15.04 so graph is clean). See the duplicate line at the bottom.  
ansi in progress

**Comment by jhowardmsft** [comment]
One more screenshot - compare with conemu with no emulation from docker cli. Looks more correct (although that's not perfect either), but the vertical spacing is correct certainly which is what seems to be the predominant issue with the new code.

ansi conemu

**Comment by gabrielhartmann** [comment]
@jhowardmsft I've captured the vertical spacing issue here.  https://github.com/Azure/go-ansiterm/issues/1

One of the key benefits of this pull request is that incremental improvements to the terminal emulation can be made against the go-ansiterm package without touch Docker's code except to update the vendor script.  This is the first draft of terminal emulation and is not my only work item at the moment.  So releasing the code so more than one person can contribute seems like a good idea to me.

Given that the current implementation crashes all the time and rarely produces even approximately correct output, I don't think this vertical spacing issue should block the pull request.

**Comment by jhowardmsft** [comment]
I agree that getting this in is definitely beneficial. LGTM :+1:

**Comment by jstarks** [comment]
This change makes Ctrl-C almost worthless. Before this change, Ctrl-C was very responsive. Now, it takes many seconds before the effects of pushing Ctrl-C. To see this, try:

    docker run -it --rm busybox find /

This will run for a long time unless interrupted. Try hitting Ctrl-C and you'll see that the results continue scrolling for about 15 seconds or so.

**Comment by gabrielhartmann** [comment]
@jstarks Issues 2 and 3 from https://github.com/Azure/go-ansiterm address the Ctrl-C problem.  We have agreed this shouldn't block the PR in an offline discussion.

@icecrime @tiborvass @crosbymichael:

Is there anything left blocking this PR?  It would be nice to get this in pretty soon.

**Comment by ahmetalpbalkan** [comment]
Looks like this PR is stalling for no reason.

Ping @docker/core-maintainers, this PR improves the ANSI emulation in Windows CLI a great deal and extracts all the Windows related code to a separate pkg, then vendors it. PTAL.

File: pkg/term/term_windows.go [hunk 2]
```diff
@@ -1,12 +1,16 @@
...
 import (
+	"fmt"
 	"io"
 	"os"
+	"os/signal"
 
 	"github.com/Sirupsen/logrus"
-	"github.com/docker/docker/pkg/term/winconsole"
+	"github.com/docker/docker/pkg/term/windows"
+	"github.com/docker/docker/vendor/src/github.com/Azure/go-ansiterm/winterm"
```
Comment (cpuguy83) [inline pkg/term/term_windows.go]: This should just be `github.com/Azure/go-ansiterm/winterm`

Comment (gabrielhartmann) [inline pkg/term/term_windows.go]: Wouldn't this directly reference the Azure/go-ansiterm repo?  Shouldn't the references go through the vendor version in the Docker repo?  Is there some other mechanism I'm unaware of which intercepts these references?  Do you mean only this file should reference the Azure repo?

I'm ok with everything directly referencing the Azure repo, but I thought the Docker team preferred snapshots of vendor code tied to particular releases/tags/commits.

Comment (cpuguy83) [inline pkg/term/term_windows.go]: Yes, the `GOPATH` on the build env is set to `/go:/go/src/github.com/docker/docker/vendor`

Comment (ahmetalpbalkan) [inline pkg/term/term_windows.go]: @gabrielhartmann the `github.com/Sirupsen/logrus` above is vendored as well. as @cpuguy83 said, the path of vendored packages is added to GOPATH while building the docker binary and thus we don't need to have the full pkg path.

For windows CLI , the GOPATH must look like: `GOPATH=/c/gopath:/c/gopath/src/github.com/docker/docker/vendor` ([you can find more here](https://ahmetalpbalkan.com/blog/compiling-docker-cli-on-windows/))

Comment (gabrielhartmann) [inline pkg/term/term_windows.go]: @cpuguy83 @ahmetalpbalkan Fixed package references and tested.

**Comment by ahmetalpbalkan** [comment]
@cpuguy83 thanks for reviewing. Looks like the commit is updated now. Can we have more core-maintainers looking at this please? The PR has approached to the 1-month mark. :smile:

**Comment by crosbymichael** [comment]
LGTM

But we should to a follow up PR to change the `var logger *logrus.Logger` var to just use the logrus functions on the package.

**Comment by jstarks** [comment]
With this change, docker client can no longer start on Nano Server because of an unnecessary call to the missing GetCurrentConsoleFont (I don't see why GetWinsize needs to return font dimensions on Windows). I've got a fix in Microsoft/docker@8652dd352586b2ed09ed607864782c85ed9de4a2, but it can't be merged because it changes the vendored go-ansiterm.

Could one of @gabrielhartmann, @ahmetalpbalkan, @brendandixon take care of this before this gets merged?

**Comment by ahmetalpbalkan** [comment]
@jstarks can you please send the PR to go-ansiterm? Then we can fix term_windows.go & re-vendor the pkg here.

**Comment by tiborvass** [comment]
Per @jhowardmsft request

**Pull request closed without merging**